wl.1
.Dd 31/07/15
.Dt wl 1
.Os PGTL
.Sh NAME
.Nm wl
.Nd Wang-Landau sampling on a 2D Ising square lattice with nearest neighbours interaction
wrapped on a torus.
.Sh SYNOPSIS
.Nm
.Op Fl g Ar double
.Op Fl s Ar double
.Op Fl t Ar double
.Op Fl l Ar unsigned int
.Op Fl f Ar double
.Op Fl r Ar unsigned int
.Sh DESCRIPTION
.Pp
All quantities are expressed in intrinsic units.
.Bl -tag -width -indent
.It Fl g -mod-factor Ar double
specifies initial modification factor for the Density of States. Must be a floating-point number larger than 1. The default value is e^1 ~ 2.718 Tip: too large a value will increase significantly the simulation time.
.It Fl s -shrink-factor Ar double
specifies the shrinking exponent for the update of the modification factor. Must be a floating-point value between 0 and 1. The default value is 0.5. Tip: a value too close to 1 will increase significantly the simulation time. A value too close to 0 will lead to very large inaccuracies in the resulting Density of States.
.It Fl t -mod-factor-threshold Ar double
specifies specifies the threshold value for the modification factor. Must be a floating-point value between 1 and the modification factor (see -g). The default value is 1 + 1e-8.
.It Fl l -edge-length Ar unsigned int
specifies the length of a single edge of the square simulation box. Values smaller than 2 are not accepted (and do not make sense anyway). Values larger than the implementation-defined maximum (can be modified in the CMakeLists.txt configuration file before compiling from source). The default value is 16.
.It Fl f -flatness-threshold Ar double
specifies a fraction of the average value of the histogram of visited energies to be used as a threshold for the flatness condition. The default value is 0.9. If ANY channel contains a value which is lower than the threshold the histogram will be considered NOT flat. Tip: too high a value may increase significantly the simulation time while too low a value will lead to great loss of accuracy in the resulting Density of States.
.It Fl r -seed Ar unsigned int
specifies seed to be fed to the pseudo-random number generator. The default value is 0, which makes the program use the default built-in seed of the pseudo-random number generator. Tip: the prng can be modified in the CMakeLists.txt configuration file before compiling from source.
.El
.Pp
.Sh FILES
.Bl -tag -width -compact
.It Pa input.txt
The input values and other sensible data is stored in this file for reproducibility and analysis.
.It Pa dos_formatted.txt
The output of the program is stored in this file as a formatted list of values. The format is
	i	e	g	log(g)
where 'i' is the C-style index of the array that used to contain the values, 'e' is the energy value associated with such index, 'g' is the value of the Density of States at the corresponding energy and 'log(g)' is its natural logarithm.
.It Pa log.txt
Contains diagnostic information on the simulation, such as the total number of performed Monte Carlo steps and their acceptance ratio.
.El
\".Sh SEE ALSO
\".Xr a 1


src/wl.cpp
//
//  main.cpp
//  wanglandau
//
//  Created by Gianluca Salvato on 18/07/15.
//  Copyright (c) 2015 Gianluca Salvato. All rights reserved.
//

#include <iostream>
#include <fstream>

#include "config.h"

#ifndef QUIET
#	define mesg(A) std::cout << A << std::endl
#else
#	define mesg(A) do {} while(0)
#endif

#include "pgtl/Algorithm/WangLandau.h"
#include "options.h"

char man_page[PATH_MAX] = "../wl.1";
_options_ cgoOptions;

using std::cin;
using std::cout;
using std::endl;

#define stringify(A) stringify_(A)
#define stringify_(A) #A

using System = PGTL::HypercubicIsingSystem<DIMENSIONALITY>;
using State = PGTL::State<System, PGTL::tags::State::NearestNeighboursInteraction>;

using Algorithm = PGTL::WangLandauAlgorithm<System, PGTL::names::Algorithm::Strategies::default_strategy>;

/*! Simulation parameters */
struct Parameters {
	Algorithm::Parameters algorithm;
	System::Size edge;
	prng_seed_type seed;
};

/*! Prints input fed to the program */
void print_input(Parameters const &);

/*! Prints output to the program */
void print_output(Algorithm const &);

int main(int argc, char * const argv[]) {

	cgoParse(argc, argv);

	Parameters params{
		.algorithm = {	.mf_log				= std::log(cgoOptions.f),
						.mf_log_shrink		= cgoOptions.f_shrink,
						.mf_log_threshold	= std::log(cgoOptions.f_threshold),
						.flatness_threshold	= cgoOptions.flatness_threshold },
		.edge = cgoOptions.edge,
		.seed = cgoOptions.seed
	};

	PRNG g{};

	Algorithm wl{System{params.edge}, params.algorithm, g};

	print_input(params);

	wl.sweep(10000);

	print_output(wl);


	return 0;
}

/*! Prints input fed to the program */
void print_input(Parameters const &params) {
	std::ofstream input_datafile("input.txt");

	input_datafile << "-- Wang-Landau Sampling --" << endl				<< endl
	<< "system_type: "			<< "Ising"								<< endl
	<< "geometry: "				<< "Hypercubic"							<< endl
	<< "dimensionality: "		<< DIMENSIONALITY						<< endl
	<< "edge_length: "			<< params.edge							<< endl
	<< ""																<< endl
	<< "mod_factor_log(f): "	<< params.algorithm.mf_log				<< endl
	<< "f_shrink_parameter: "	<< params.algorithm.mf_log_shrink		<< endl
	<< "f_threshold_log: "		<< params.algorithm.mf_log_threshold	<< endl
	<< "flatness_threshold: "	<< params.algorithm.flatness_threshold	<< endl
	<< ""																<< endl
	<< "prng: "					<< stringify(PRNG)						<< endl
	<< "prng_seed: "			<< params.seed							<< endl;

	input_datafile.close();
}

#include <functional>
#include <limits>

/*! Prints output to the program */
void print_output(Algorithm const &a) {
	std::ofstream	dos_datafile("dos_formatted.txt"),
					log_datafile("log.txt");

		// Print basic information

	log_datafile << "-- Wang-Landau Sampling --"	<< endl	<< endl
	<< "monte_carlo_steps: "	<< a.moved_total			<< endl
	<< "monte_carlo_sweeps: "	<< a.sweeps					<< endl
	<< "accepted_steps: "		<< a.accepted_total			<< endl
	<< "acceptance_ratio: "		<< a.acceptance_ratio()		<< endl;

		// Print actual data in dos_datafile

	Algorithm::EnergyChannels const &energy = a.energy;
	Algorithm::DoSLogHistogram const &dos_log = a.dos_log;
	Algorithm::DoSHistogram dos(dos_log.size());

	auto f = [](Algorithm::DoSLog const &x) { return std::exp(x); };

	std::transform(dos_log.begin(), dos_log.end(),
				   dos.begin(), f);

	dos_datafile.precision(std::numeric_limits<Algorithm::DoS>::digits10);
	for (Algorithm::Index i = 0; i < a.dos_log.size(); i++) {
		dos_datafile	<< i			<< "\t"
						<< energy[i]	<< "\t"
						<< dos[i]		<< "\t"
						<< dos_log[i]	<< endl;
	}

	dos_datafile.close();
	log_datafile.close();
}

#undef DIMENSIONALITY
#undef MAX_SITES
#undef PRNG

#undef stringify
#undef stringify_

#undef mesg
